### PR TITLE
PP-129 OPDS 2 with acquisition links

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
This PR should be reviewed in tandem with the [Feed Refactor PR](https://github.com/ThePalaceProject/circulation/pull/1308).
## Description
Acquisition links for the OPDS 2 feeds as per the [De Marque document](https://docs.google.com/document/d/1tRy8uh4ckC9WDpjKXLedg87FDwsT4syv-Jvbp5dFFMk/edit).
This PR also adds the ability for certain feeds to be served as either XML or JSON based on the requests `content-type`.
This PR was also used to complete the OPDS2 serializer which was left out of #1308 .

Example
```
      {
          "href": "http://localhost:6500/VT0014/works/Overdrive%20ID/0e20dfec-501f-4b5d-926f-dd712aa39160/borrow",
          "rel": "http://opds-spec.org/acquisition/borrow",
          "type": "application/atom+xml;type=entry;profile=opds-catalog",
          "properties": {
            "availability": {
              "state": "available"
            },
            "indirectAcquisition": [
              {
                "type": "application/vnd.adobe.adept+xml",
                "child": [
                  {
                    "type": "application/pdf"
                  }
                ]
              },
              {
                "type": "application/vnd.overdrive.circulation.api+json;profile=ebook"
              },
              {
                "type": "application/atom+xml;type=entry;profile=opds-catalog",
                "child": [
                  {
                    "type": "text/html;profile=\"http://librarysimplified.org/terms/profiles/streaming-media\""
                  }
                ]
              }
            ]
          }
        }
```

#### Missing
The `authenticate` property has not been added, since we are currently providing the authentication document link as part of the feed links. Wasn't sure is we wanted to change that behaviour.

## Motivation and Context
[JIRA-129](https://ebce-lyrasis.atlassian.net/browse/PP-129) and [JIRA-78](https://ebce-lyrasis.atlassian.net/browse/PP-78)

## How Has This Been Tested?
Manually checked 

## Checklist

- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by jonathangreen** [comment]
@RishiDiwanTT The first thing I was trying to do with this PR is to load the OPDS2 version of the feed in my test environment to see what the serialized OPDS2 looks like. Can you give me some pointers how I can get an OPDS2 response?

**Comment by RishiDiwanTT** [comment]
Currently only the /loans, /borrow and /feed APIs are activated for opds2.
They can be seen by adding a content type : application/opds+json header to
the GET request.

**Comment by jonathangreen** [comment]
Typically the [content-type header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) indicates the type of the data included in the request. I'd expect to do content negotiation like this via an [accept header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept) included on the request, which would cause the content-type header to be included in the response.

**Comment by RishiDiwanTT** [comment]
That makes sense, and is a very small change.

**Comment by jonathangreen** [comment]
Okay. So I pointed this branch at an existing DB with some books in it, and then I ran the resulting OPDS2 feed through our OPDS2 validator `/bin/opds2_schema_validate` and our feed is failing our own validation. So i think we need to get that sorted out to start with.

**Comment by RishiDiwanTT** [comment]
> Okay. So I pointed this branch at an existing DB with some books in it, and then I ran the resulting OPDS2 feed through our OPDS2 validator `/bin/opds2_schema_validate` and our feed is failing our own validation. So i think we need to get that sorted out to start with.

This ended up being trivial, changes required were 2
- Use the "Accept" header instead of the "Content-Type" header
- The "audience" scheme should be an URL always

Apart from that the actual validator code was broken, I guess somewhere along the way the OPDSImporter code changed a bit and the validator did not reflect those changes.

**Comment by jonathangreen** [comment]
From the [comment](https://github.com/ThePalaceProject/circulation/pull/1308#discussion_r1310613869) on #1308:

> [jonathangreen](https://github.com/jonathangreen) [Aug 30, 2023](https://github.com/ThePalaceProject/circulation/pull/1308#discussion_r1310613869)
Can we rename the package (directory) for this code as well? Right now its feed_protocol, I'd suggest feed.

> [RishiDiwanTT](https://github.com/RishiDiwanTT) [Sep 1, 2023](https://github.com/ThePalaceProject/circulation/pull/1308#discussion_r1312717382)
This will cause all kinds of unpleasantness while rebasing with https://github.com/ThePalaceProject/circulation/pull/1340 . I will do this once that PR has been merged with this.

Moving this to a comment here so we don't forget the package needs to be renamed.

**Comment by RishiDiwanTT** [comment]
> From the [comment](https://github.com/ThePalaceProject/circulation/pull/1308#discussion_r1310613869) on #1308:
> 
> > [jonathangreen](https://github.com/jonathangreen) [Aug 30, 2023](https://github.com/ThePalaceProject/circulation/pull/1308#discussion_r1310613869)
> > Can we rename the package (directory) for this code as well? Right now its feed_protocol, I'd suggest feed.
> 
> > [RishiDiwanTT](https://github.com/RishiDiwanTT) [Sep 1, 2023](https://github.com/ThePalaceProject/circulation/pull/1308#discussion_r1312717382)
> > This will cause all kinds of unpleasantness while rebasing with #1340 . I will do this once that PR has been merged with this.
> 
> Moving this to a comment here so we don't forget the package needs to be renamed.

Do we have anymore feedback for this PR or should I merge it into the refactor branch?

**Comment by jonathangreen** [comment]
@RishiDiwanTT I'm taking another look over this PR now. I'll post a comment once I'm done.

File: tests/api/feed_protocol/test_opds_acquisition_feed.py
Comment (jonathangreen) [inline tests/api/feed_protocol/test_opds_acquisition_feed.py:63]: Typo in comment

File: api/controller.py
Comment (jonathangreen) [inline api/controller.py:964]: I think we need to do something a bit more sophisticated then this to properly support [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation). Some more detail here: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

This is a valid accept header:
`Accept: application/atom+xml;profile=opds-catalog;kind=acquisition, application/opds+json;q=0.9, */*;q=0.8`

Which would indicate that the client wants OPDS 1.x response. But we would return an OPDS 2.x response.

It seems like flask provides some parsing of accept headers that we should be using in [Request.accept_mimetypes](https://flask.palletsprojects.com/en/2.3.x/api/#flask.Request.accept_mimetypes), rather then using the header directly.

File: tests/api/feed_protocol/test_library_annotator.py
Comment (jonathangreen) [inline tests/api/feed_protocol/test_library_annotator.py:935]: Why does this test come out?

**Review by jonathangreen — COMMENTED** [review]
@RishiDiwanTT I made a few more comments here

File: core/feed_protocol/acquisition.py
Comment (RishiDiwanTT) [reply core/feed_protocol/acquisition.py:597]: I thought of doing that, but I got lazy... told you it wasn't fixed.

File: tests/api/feed_protocol/test_library_annotator.py
Comment (RishiDiwanTT) [reply tests/api/feed_protocol/test_library_annotator.py:935]: This code was removed from the annotator, and [added into the controller](https://github.com/ThePalaceProject/circulation/pull/1340/files#diff-944bc0ea08885efcf096971a642422c2f5ea80605fd0b10618c9e6a738d7a837R1470), since the annotator no longer creates the serialized response.
The tests in `tests/api/test_controller_loan.py` already test for that functionality.

File: core/feed_protocol/acquisition.py
Comment (jonathangreen) [inline core/feed_protocol/acquisition.py:593]: It looks like this new case isn't covered by a test, and it seems like the original code doesn't return a problem detail in this case, but converts the `OPDSMessage` into an error response?

Can you write up a test for this and verify that we aren't changing the behavior of the single item loans feed here?

File: core/feed_protocol/opds.py
Comment (jonathangreen) [inline core/feed_protocol/opds.py:16]: This is closer, but I think we need to write some tests around this function. 

I did a little bit of testing, and I don't think iterating over a `MIMEAccept` puts the results in priority order. It uses the order that the Accept header was sent in. It also returns a `Tuple[str, int]` mime / quality. 

I think we probably want to use the [best_match method](https://werkzeug.palletsprojects.com/en/2.0.x/datastructures/#werkzeug.datastructures.Accept.best_match) of `MIMEAccept`, which does the kind of sorting / matching we want here.

It would be nice to see some tests that test a few cases here: 
- No mime passed in
- A single simple mime type
- A couple more complicated Accept headers like: 
  - `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition;q=0.08, application/opds+json;q=0.9`
  - `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition, application/opds+json;q=0.9`

Comment (RishiDiwanTT) [reply core/feed_protocol/opds.py:16]: Ok, I didn't quite grasp the q-value sets the priority and not the order of appearance. I'll fix this.

File: core/feed_protocol/acquisition.py
Comment (jonathangreen) [reply core/feed_protocol/acquisition.py:593]: I think either: 
- we need to handle this at the serialization step, so that we get the same XML response we used to get in the case of an OPDS 1 feed
- Or check with the mobile apps and see how returning a problem detail here instead of a feed will be handled

Since our mobile apps are the primary consumer of this feed still, we need to make sure they are in sync with any changes that get made to the responses.

Comment (RishiDiwanTT) [reply core/feed_protocol/acquisition.py:593]: We can return to the XML format easily, will that be an acceptable behaviour when the client is expecting JSON though?
Re: ProblemDetail: The same method also returns problem details [in different circumstances](https://github.com/ThePalaceProject/circulation/pull/1340/files/8c4941b94515b085a85b482a88d366642315b588..508b81a66336ff8bd846c90401df0ce490ba6391#diff-915f371a570738a5b9970eca750c31619010b101b4ca87634fcbcaf01649c078R561) so a problem detail should be already handled by the client side.

File: core/feed_protocol/opds.py
Comment (jonathangreen) [reply core/feed_protocol/opds.py:16]: Can you add a test case for these Accept header? 
- `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition;q=0.08, application/opds+json;q=0.9`
  - I'd expect an `application/opds+json` response in that case, but I believe the current implementation will response with `application/atom+xml`. 
- `Accept: application/atom+xml;profile=opds-catalog;kind=acquisition`
  - I'd expect an `application/atom+xml` response in that case, but I believe we will serve `application/opds+json`

Comment (RishiDiwanTT) [reply core/feed_protocol/opds.py:16]: Well... that was embarrassing. Seems like the MIMEAccept datatypes sorting is based FIRST on specificity THEN on q-value.
Changed the method to always sort on q-value.

Comment (jonathangreen) [reply core/feed_protocol/opds.py:16]: Is there a reason you aren't using the best_match method I mentioned above here? It would save a lot of sorting etc and lets you set a default.

Comment (RishiDiwanTT) [reply core/feed_protocol/opds.py:16]: The datatype could be a List, this is only used in a test case for now, but was done to future proof the method in case we needed to do generate feeds offline.

Comment (jonathangreen) [reply core/feed_protocol/opds.py:16]: @RishiDiwanTT I believe that taking the specificity into account is part of the usual Accept header content negotiation dance. It would be nice to be able to lean on the complex matching provided by Flask instead of trying to reimplement it ourselves.

For tests, can we just directly pass in a `MimeAccept` class, or use the (admittedly annoying) flask machinery to fake a request for testing so we can directly consume `MimeAccept`.

Comment (RishiDiwanTT) [reply core/feed_protocol/opds.py:16]: Updated. This caused one test case to change because specificity magic switched around the best match.

**Review by jonathangreen — COMMENTED** [review]
I'm still concerned here about the change in behavior of the `OPDSMessage` class as part of this PR. I don't want to hold up merging this, but I also don't want it to have a negative impact on the experience of our users consuming the OPDS 1 feed through the app.

File: core/feed_protocol/opds.py
Comment (RishiDiwanTT) [reply core/feed_protocol/opds.py]: I introduced an abstract base type to better handle the serializer protocol

File: core/feed_protocol/acquisition.py
Comment (RishiDiwanTT) [reply core/feed_protocol/acquisition.py]: Ok, I reverted to protocol specific data, OPDS1 will be XML and OPDS2 will be JSON.

**Review by jonathangreen — APPROVED** [review]
🚀 I think this is looking good @RishiDiwanTT

**Pull request merged**
